test/bpf_ut/bpftest/bpf_test.go

```diff
@@ -0,0 +1,559 @@
+		t.Skip("Set -bpf-ut-path to run BPF tests")
+	}
+
+	if err := rlimit.RemoveMemlock(); err != nil {
+		t.Log(err)
+	}
+
+	t.Run("Workload", testWorkload)
+}
+
+// common functions
+
+func loadAndRunSpec(t *testing.T, tt *unittest) {
+	elfPath := path.Join(*testPath, tt.objFilename)
+	t.Logf("Running test %s", elfPath)
+
+	spec := loadAndPrepSpec(t, elfPath)
+
+	var (
+		coll *ebpf.Collection
+		err  error
+	)
+
+	// Load the eBPF collection into the kernel
+	coll, err = ebpf.NewCollection(spec)
+
+	// Check for errors, specifically handle eBPF verifier errors separately
+	// as they provide more detailed diagnostics
+	if err != nil {
+		var ve *ebpf.VerifierError
+		if errors.As(err, &ve) {
+			t.Fatalf("verifier error: %+v", ve)
+		} else {
+			t.Fatal("loading collection:", err)
+		}
+	}
+	defer coll.Close()
+
+	// Iterate through all programs and organize them by test name and function type
+	// based on their section names matching the pattern "<test_name>/(check|jump|pktgen)"
+	programs := programSet{}
+	for progName, spec := range spec.Programs {
+		match := checkProgRegex.FindStringSubmatch(spec.SectionName)
+		if len(match) == 0 || match[1] != tt.name {
+			continue
+		}
+
+		if match[2] == "pktgen" {
+			programs.pktgenProg = coll.Programs[progName]
+		}
+		if match[2] == "jump" {
+			programs.jumpProg = coll.Programs[progName]
+		}
+		if match[2] == "check" {
+			programs.checkProg = coll.Programs[progName]
+		}
+	}
+
+	// Ensure test that has a jump program also has a check program
+	if programs.checkProg == nil {
+		t.Fatalf(
+			"File '%s' contains a jump program in section '%s' but no check program.",
+			elfPath,
+			tt.name,
+		)
+	}
+
+	// Collect bpf log events and add them as logs of the main test
+	collectBpfLogEvents(t, coll)
+	if tt.setupInUserSpace != nil {
+		tt.setupInUserSpace(t, coll)
+	}
+
+	// run the test
+	subTest(programs, coll.Maps[suiteResultMap])(t)
+}
+
+// collectBpfLogEvents sets up a ring buffer reader for the "km_log_event" eBPF map
+// and logs messages received from the BPF program. This function is a test helper
+// that creates a goroutine to continuously read and process log events.
+//
+// The function skips setup on kernel versions lower than 5.13 as the feature
+// may not be supported. It reads from a ringbuf map named "km_log_event" if present
+// in the provided collection.
+//
+// Each log event is decoded from binary format into a LogEvent structure. Messages
+// that don't start with "bpf log level" are logged to the test logger.
+//
+// Parameters:
+//   - t: The testing context used for logging and test failure reporting
+//   - coll: The eBPF collection containing loaded maps and programs
+//
+// Returns:
+// The goroutine started by this function will continue running until the ringbuff.CollectionSpec: The prepared collection specification
+// reader encounters an error (typically when the map is closed).
+func collectBpfLogEvents(t *testing.T, coll *ebpf.Collection) {
```

Comment:
some sililarities with pkg/logger/logger.go i am wondering can we abstract a utility that can be shared